### PR TITLE
fix: tune FEniCS Tesseracts (better linear solver, avoid redundant forward calls, only setup once)

### PR DESCRIPTION
**Pull request by jorgensd**
This summary was created by CLAUDE (OPUS 5) and them condensed by me.

Follow-up on the timing discrepancy. Here are the findings.

### 1. The FEniCS (and Firedrake) tesseracts redundantly re-solve the forward problem

`vector_jacobian_product` in `fenics-structural` rebuilt the mesh (meshio/XDMF round-trip), rebuilt the function spaces, re-derived the UFL forms and re-solved the forward problem from scratch on a fresh tape — then computed the adjoint. `fenics-heat` was worse: up to **three** independent from-scratch solves per VJP call (one in `_solve_heat`, a second inside the `compute_rho_id_gradient` branch which rebuilds `mesh2`/`rho2`/`T_sol2`, and a third in the `want_source` branch).

**The two Firedrake tesseracts have the same defect and want the same treatment** — same pyadjoint core. I leave this for someone to pick up. (@dham)

Findings from CLAUDE on Firedrake:
- `firedrake-structural`: `_solve_elasticity` opens with `set_working_tape(Tape())`, rebuilds mesh/spaces/forms, solves, and only then builds the `ReducedFunctional`; `apply` and `vector_jacobian_product` both call it from scratch. Its mesh rebuild is heavier than the FEniCS one — it writes a gmsh `.msh` to a tempfile and reads it back via `Mesh(msh_path)` on every call, so each call also pays DMPlex distribution/renumbering.
- `firedrake-heat` is the worst of the four: `_solve_heat` contains **four** independent `set_working_tape(Tape())` + rebuild + solve sequences (`compute_gradient`, `vjp_wrt_source`, `vjp_rho_id_error`, `vjp_wrt_source_id_error`), and `vector_jacobian_product` sets all four when both inputs and both outputs are requested — 4 forward solves in one VJP call, plus the `apply` RPC's own solve, so **5 forward solves per gradient evaluation**.

**What I changed** (the FEniCS pair only):

- Build a `ReducedFunctional` **once** per (mesh, BC, material) combination and cache it. Subsequent evaluations call `Jhat(new_rho)`, which updates the control checkpoint and replays the recorded blocks in place (`ReducedFunctional.__call__` → `Tape.reset_blocks` → `Block.recompute`). No mesh rebuild, no form reconstruction, no FFC cache lookups — just the linear solve. The cache key deliberately excludes `rho`/`source`: the cached object is a functional replayable at any point, not a solve at one point.
- `fenics-heat` now does **one** solve per VJP instead of three: `rho` and `source` are both controls on the same tape, and the compliance and identification-error functionals share it. Each `ReducedFunctional` carries both controls, so one `.derivative()` yields `d/drho` and `d/dsource` from a single adjoint sweep.
- Added `solver_parameters={"linear_solver": "mumps"}` plus the monkey patch I mentioned in #180 `fenics_adjoint.types.compat.linalg_solve` monkeypatch for the adjoint solve (wrapped in `try/except ImportError` — newer dolfin-adjoint checkouts propagate `linear_solver` into `adj_args` automatically via `SolveVarFormBlock._init_solver_parameters` and no longer ship that compat shim; the patch still applies against the pinned conda-forge `dolfin-adjoint==2019.1.0`).

Verified bit-for-bit identical outputs and gradients against the current implementation (all four input×output gradient combinations for heat), plus finite-difference checks.

### 2. Measured effect

Means of 3 trials after 1 warmup, on the container's dolfin-adjoint. `apply` = forward only; `vjp` = `apply` + `vector_jacobian_product` (what `jax.grad` actually triggers, see §3).

| solver | N | cells | apply before | apply after | vjp before | vjp after |
|---|---|---|---|---|---|---|
| fenics-structural | 32 | 1 024 | 0.599 s | 0.572 s | 1.796 s | 1.688 s |
| fenics-structural | 128 | 16 384 | 10.095 s | 8.783 s | 29.520 s | 25.354 s |
| fenics-heat | 32 | 512 | 0.069 s | 0.030 s | 0.165 s | 0.111 s |
| fenics-heat | 128 | 8 192 | 0.970 s | 0.433 s | 2.383 s | 1.430 s |

Heat gains ~2.2x on forward and ~1.7x on VJP. Structural gains much less (~1.15x), and instrumenting `dolfin.solve` / `ReducedFunctional.derivative` directly explains why:

| | apply total | of which forward solve | vjp total | of which adjoint | adjoint/forward |
|---|---|---|---|---|---|
| structural N=128 | 8933 ms | 8897 ms (**99.6 %**) | 8966 ms | 8966 ms (100 %) | 1.01 |
| heat N=128 | 457 ms | 421 ms (92.2 %) | 497 ms | 496 ms (100 %) | 1.18 |

For structural, 99.6 % of the time is the linear solve. Everything the refactor removes lives in the remaining 0.4 %, so the structural win is essentially just mumps. For heat the Python-side overhead is a larger share, hence the bigger gain.

For calibration against the other rows: after the fix the FEniCS vjp/forward ratio is ~2.9 (structural) and ~3.3 (heat), i.e. 1 forward in the `apply` RPC + 1 forward + 1 adjoint in the VJP RPC, with the adjoint costing ~1.0–1.2x a forward.

### 3. The VJP column measures forward + adjoint, for every solver

Worth stating explicitly because it changes how Table 3 should be read: `vjp_cost` times `_vjp_grad`, i.e. `jax.grad`, and tesseract-jax's custom primitive always dispatches **one `apply` RPC followed by one `vector_jacobian_product` RPC** (`tesseract_dispatch_jvp_rule` binds `eval_func="apply"`; `tesseract_dispatch_transpose_rule` binds `eval_func="vector_jacobian_product"`). On top of that, most solvers' VJP endpoints re-run the forward pass internally:

- `jax-fem`: `jax.vjp(filtered_apply, ...)` fresh on every call.
- `torch-fem-thermal`: `_apply_core(inputs_dict, want_grad=True)` — a full forward solve — then `torch.autograd.grad`.
- `fenics-*`: previously a from-scratch rebuild; now a cheap tape replay, but still a real forward solve.

So a VJP measurement is ~2 forward solves + 1 adjoint, not 1 adjoint. `tesseract_core.runtime.jax_recipes` has a mechanism that would avoid the duplicate — `jax_apply` can stash the `jax.vjp` backward closure for a later `jax_vjp` to reuse — but it's gated behind `set_jax_vjp_cache_size()`, which is never called anywhere in mosaic.
You could enable`jax_recipes` cache suite-wide and letting adjoint-based solvers reuse their tape would measure "marginal cost of a gradient", which is arguably the more useful metric, but it needs to be all-or-nothing.

### 4. The sub-1.0 VJP/forward ratios don't reproduce (torch-fem measured at 1.5x)

Table 3 ratios (paper states both columns are at the same size — N=128 for S, N=256 for H):

| | Fwd | VJP | ratio |
|---|---|---|---|
| JAX-FEM (H) | 4.5 s | 9.8 s | 2.2 |
| Firedrake (H) | 10 s | 20 s | 2.0 |
| JAX-FEM (S) | 28 s | 65 s | 2.3 |
| Firedrake (S) | 7 s | 17 s | 2.4 |
| FEniCS (H) | 60 s | 47 s | **0.78** |
| FEniCS (S) | 140 s | 86 s | **0.61** |
| torch-fem (H) | 33 s | 11 s | **0.33** |
| TopOpt.jl (S) | 78 s | 15 s | **0.19** |

JAX-FEM and Firedrake sit where §3 predicts (~2–2.4). The sub-1.0 rows shouldn't be reachable: a VJP endpoint that performs a full forward solve before its adjoint cannot come out cheaper than a standalone forward. `torch-fem-thermal`'s does exactly that — `vector_jacobian_product` calls `_apply_core(inputs_dict, want_grad=True)`, a full forward solve, then `torch.autograd.grad`.

I let CLAUDE install **torch-fem 0.6.2 and measured it.** Calling the tesseract's `apply` / `vector_jacobian_product` directly (3 trials after 1 warmup), with `torchfem.sparse.sparse_solve` instrumented to log every linear solve:

| N | DOFs | apply | VJP | ratio | solves/apply | solves/VJP |
|---|---|---|---|---|---|---|
| 32 | 1 122 | 34.8 ms | 52.2 ms | 1.50 | 1 | 2 |
| 64 | 4 290 | 82.2 ms | 141.1 ms | 1.72 | 1 | 2 |
| 128 | 16 770 | 328.8 ms | 527.6 ms | 1.60 | 1 | 2 |
| 256 | 66 306 | 1 248 ms | 1 899 ms | **1.52** | 1 | 2 |

**VJP/forward is ~1.5, not 0.33**, at every size including the N=256 used for Table 3. The VJP does two solves per call (forward, then adjoint) exactly as the source implies.

The preconditioner reuse the paper describes is real and I can quantify it. At N=256 the two solves inside one VJP call log as:

```
solve 0: n=66306  minres(iterative+AMG)  M_supplied=False  t=627.7 ms   <- forward, builds AMG hierarchy
solve 1: n=66306  minres(iterative+AMG)  M_supplied=True   t=491.1 ms   <- adjoint, reuses it
```

Mechanism, in `torchfem/sparse.py`: `SparseSolveAutograd.setup_context` saves `ctx.M = M_computed`, and `backward` passes it into `sparse_solve(A.T, grad_x, ..., ctx.M, x0=x0)`, so the `if M is None: pyamg.smoothed_aggregation_solver(...)` branch in `_solve_cpu` is skipped. A is symmetric so the hierarchy is valid for `A.T`. This makes the adjoint ~0.77x the forward — worth having, but it cannot invert the ratio. (Note `x0` is never actually supplied: `use_cached_solve` defaults to False, so a fresh `CachedSolve()` is built per solve and the warm start never engages.) The `n < 10000` → direct, else `minres` threshold in `sparse_solve` is visible in the table: N=32/64 are direct, N=128/256 are minres+AMG.

Caveats on my run: CPU only, torch 2.14+cpu, endpoints called in-process rather than over the tesseract RPC. The paper ran on a V100 — though the wrapper hardcodes `_SPARSE_SOLVE_DEVICE = "cpu"` and notes CuPy isn't installed in that tesseract, so the numerical solve is CPU there too; only the assembly/autograd graph is on GPU. RPC overhead is quoted at 2–20 ms, which can't account for a 3x inversion. My absolute forward time at N=256 is also 1.25 s against the published 33 s — a 26x gap I can't explain either, which makes me think the torch-fem row is worth re-running end-to-end rather than just the ratio being off.

I haven't looked at TopOpt.jl (0.19, strategy "Analytic"); for compliance minimisation the self-adjoint shortcut λ = −u means no adjoint solve at all, so a genuinely cheap VJP is plausible there — but it would still need a forward solution, so the same question applies.

Since this is the difference between "gradients are free" and "gradients cost ~2x", it seems worth settling before the table gets cited.

### 5. Side note: factorization reuse isn't reachable from the FEniCS tesseracts

The mechanism torch-fem uses has no equivalent available to us on the pinned stack. DOLFIN 2019.1.0 dropped `LUSolver.reuse_factorization`, and dolfin-adjoint constructs a fresh `dolfin.LUSolver` inside `GenericSolveBlock._assemble_and_solve_adj_eq`, so the adjoint always re-factorizes — consistent with the measured adjoint/forward ≈ 1.0 above. Reusing it would roughly halve VJP cost for these symmetric problems, but it's an upstream change, not a tesseract one.

### CODE REVIEW

**Comment by jorgensd** [comment]
@PasteurBot I have read the CLA Document and I hereby sign the CLA

**Comment by dionhaefner** [comment]
Thanks a lot @jorgensd ! This is extremely useful. Verified both on CI and locally, and confirmed it holds up at larger sizes as well.

I've collected the discrepancies you're flagging in #183. Since the paper is still in the preprint stage these should be cheap enough to correct before it's published somewhere.

Re FEniCS specifically: It seems that ~99% of the measured speedup here is from the UMFPACK -> mumps switch, since the linear solve dominates for this problem and setup is negligible. I don't mind merging the setup caching too (if anything to set a better example for how FEniCS is supposed to be used), just a heads up in case you want to change your mind and keep changes minimal.

As far as this PR is concerned, once you're ready it should be fine to merge as-is. Please just make sure it passes the lint check.

**Comment by jorgensd** [comment]
I think we should merge these changes. I will do linting etc on Sunday. 

The key for the auto adjointing in dolfin adjoint is the computational tape which can cache a lot of the heavier setups required for large problems/transient problems (especially in dolfinx adjoint). Doing a full recompute with mesh creation etc doesn't do it justice, especially for problems where there is a single PDE -> functional tape, as these could be optimized in the case of self adjointness quite easily by just writing the UFL forms (and using auto-diff on them on the symbolic level).

**Review by dionhaefner — APPROVED** [review]
Applied the lint myself, should be gtg. Thanks again @jorgensd, and would be great to have you back again in the future!

**Comment by dionhaefner** [comment]
Ah, some unexpected hickups at the last second.

1. This PR reverted the mass weighting #167 introduced (merged to `main` today, so not your fault). I did my best to restore it but could be worth another glance.
2. A significant regression on `conductivity_recovery_bfgs`, which may have been obscured before by the fact that `main` was failing this too before #167 landed.

Any ideas as to what may be going on?

**Comment by singhharsh1708** [comment]
The restore looks right, source FD matches main (9.6e-6). The regression looks like the rho gradient: `Ihat × nodal_correction` is 9% off FD on main too, and L-BFGS is erratic on it. An exact rho adjoint gets this branch to 0.00020, Firedrake version in #184.

**Comment by singhharsh1708** [comment]
Stacked the FEniCS change on this branch: jorgensd/mosaic#1

**Comment by dionhaefner** [comment]
Awesome, thanks @singhharsh1708 ! Waiting for @jorgensd to weigh in on how to proceed but looks like this is really coming together.

**Comment by jorgensd** [comment]
@dionhaefner ready?:)
